Lightning integration

szmazurek opened the pull request:
Fixes #

## Proposed Changes
- added lightinign-based logic to the training and inference managers
- removed bugs found on the way
- conducted testing and debugging - still, some edge cases remain that are causing some tests to fail, altoguht they semm quicly fixable

## Checklist

- [ ] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [ ] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [ ] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [ ] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [ ] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [ ] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [ ] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].
- [ ] The `logging` library is being used and no `print` statements are left.

szmazurek committed:
training manager lightning integration

szmazurek committed:
Update train manager split

szmazurek committed:
Remove device param from entrypoint

szmazurek committed:
Lightning integration

szmazurek committed:
Fix bug with lacking spacing keyword for distance metrics

szmazurek committed:
Preparation for sdnet bug removal, grad clip type removal

szmazurek committed:
Debugging

- removed unnecessary parameters from test
- marked sdnet test as skip
- solved the problem of incrrect metric tensor shape when predicitons
  became negative

szmazurek committed:
Fixed the problem with device mismatch

szmazurek committed:
Bugfix

szmazurek committed:
Correct scheduler intialization

szmazurek committed:
Correct redundant device param

szmazurek committed:
Correct loss errors

szmazurek committed:
Ensure param present

szmazurek committed:
Fix histo test WIP

szmazurek committed:
Fix histo test

szmazurek committed:
Last bug fixes

szmazurek committed:
Added newly added params to the sample config

szmazurek committed:
Spelling

szmazurek committed:
Spelling error

sarthakpati commented on GANDLF/metrics/metric_calculators.py:
```diff
@@ -13,7 +14,9 @@ def __init__(self, params: dict):
-    def _process_metric_value(self, metric_value: torch.Tensor):
+    def _process_metric_value(self, metric_value: Union[torch.Tensor, float]):
+        if isinstance(metric_value, float):
+            return metric_value
```
Should the bottom `if` be `ifelse`?

szmazurek replied:
Could be but, not necessairly, each statement will be checked anyway - this method will return something, if we missed any case then it means it is a bug.

sarthakpati commented on samples/config_all_options.yaml:
```diff
@@ -95,7 +95,7 @@ num_epochs: 100
 # Set the batch size
 batch_size: 1
-# gradient clip : norm, value, agc
+# gradient clip : norm, value
```
Since [Lightning has its own gradient clipping](https://lightning.ai/docs/pytorch/stable/common/optimization.html#gradient-clipping), Do we need our custom [gradient clipping code](https://github.com/mlcommons/GaNDLF/tree/master/GANDLF/grad_clipping) anymore?

szmazurek replied:
Not really, should I remove it too?

szmazurek replied:
Removed

sarthakpati commented on samples/config_all_options.yaml:
```diff
@@ -304,3 +304,7 @@ q_samples_per_volume: 5
 # used for debugging
 q_verbose: False
+
+accelerator: "cpu" # options: "cpu", "gpu", "auto", "auto" selects GPU if available
```
These should be part of docs.

szmazurek replied:
Right, will add

szmazurek replied:
Added

sarthakpati commented on testing/test_full.py:
```diff
@@ -2913,6 +2910,7 @@ def test_generic_deploy_docker():
     print("passed")
 
 
+@pytest.mark.skip("Differential privacy not implemented")
```
??? Does Lightning not support Opacus?

szmazurek replied:
It does, but I purposefully skipped during implementation now, as I wanted to have base implementation working now. Do we also need it for the current port?

sarthakpati replied:
It would be good to have that included as well, since there are active experiments that utilize this. Could be another benchmarking test on our (primarily my) end as well.

szmazurek replied:
added

szmazurek committed:
Opacus abstraction

szmazurek committed:
Opacus implementation

szmazurek committed:
Setting dataloader for DP corrected

szmazurek committed:
Differential privacy support

szmazurek committed:
Fix bug

szmazurek committed:
Test fully working

szmazurek committed:
Remove logging test bug

szmazurek committed:
Entrypoint correction for new args

szmazurek committed:
Remove old code, add docs

szmazurek committed:
Legacy code and spelling

sarthakpati commented on GANDLF/_graveyard/README.md:
I would recommend renaming the `_graveyard` submodule to `_deprecated`

szmazurek replied:
ok, good idea

szmazurek committed:
Restore files for openfl

szmazurek committed:
Remove the deprecated due to OpenFL fails

szmazurek committed:
Lr finding working

szmazurek committed:
Batch size finder working

szmazurek committed:
Training batch size and lr finders working

szmazurek committed:
Inference automatic batch size finder

szmazurek committed:
Final bugfixing

szmazurek committed:
Spelling correction

szmazurek committed:
Correcting missing logger headers

szmazurek committed:
Remove queue reinit when automatic lr/batch size finding

szmazurek committed:
Add the docs and new config options

szmazurek committed:
Add device deprecation

sarthakpati commented on GANDLF/compute/generic.py:
```diff
@@ -7,42 +7,49 @@
 
 class TrainingSubsetDataParser(AbstractSubsetDataParser):
-    def create_subset_dataloader(self) -> DataLoader:
+    def create_subset_dataset(self):
```
Return typing hint could be useful.

szmazurek replied:
In this case, the interpreter is able to define it automatically (as a return type from ImagesFromDataframe), so it is appearing in type hints anyway and mypy is detecting it. Do we want to write it explicitly if that is the case?

sarthakpati replied:
Ah okay. I think it could be useful (from a coding perspective).

szmazurek replied:
cool, so I will scan the module and it it whenever missing

sarthakpati commented on GANDLF/compute/generic.py:
```diff
@@ -64,31 +71,44 @@ def create_subset_dataloader(self) -> DataLoader:
 
 class ValidationSubsetDataParser(AbstractSubsetDataParser):
-    def create_subset_dataloader(self) -> DataLoader:
+    def create_subset_dataset(self):
```
Return typing hint could be useful.

sarthakpati commented on GANDLF/compute/generic.py:
```diff
@@ -64,31 +71,44 @@ def create_subset_dataloader(self) -> DataLoader:
-        # hack to cope with the get_testing_loader() function
-        self.parameters_dict["testing_data"] = self.subset_csv_path
-        return get_testing_loader(self.parameters_dict)
+    def create_subset_dataset(self):
```
Return typing hint could be useful.

sarthakpati commented on GANDLF/inference_manager.py:
```diff
@@ -112,17 +111,15 @@ def InferenceManager(
-                else:
-                    LightningTuner(trainer).scale_batch_size(
-                        lightning_module, datamodule=datamodule, method="predict"
+                        "Auto batch size find is not supported in inference. Dataloader batch size is always 1."
```
If `auto_batch_size_find` is not supporter, why is the check only for `["path", "histo"]`? Should it not be for `rad` as well?

szmazurek replied:
That's correct! Missed this one

szmazurek committed:
Profiling with lightning

szmazurek committed:
Spelling config correction

szmazurek committed:
Add missing docstring

szmazurek committed:
Remove unecessary warning

szmazurek committed:
workflow debugging

szmazurek committed:
workflow debugging

szmazurek committed:
Removing OOM error on regression 3D test

No idea where it came from, had to just downsize the patch size. It is
caused during evaluation of efficient net, and not always
